### PR TITLE
[ESP32] Add ESP-IDF 3.3 support

### PR DESCRIPTION
**Pull request by erjiaqing**
This PR cherry picks [`f70c83b`](https://github.com/openweave/openweave-core/commit/f70c83b82b6202e36b5a782abe524edd4a997480) to master.

After patched ESP-IDF 3.2, these updates is required by ESP-IDF 3.3

- Update `component.mk` to build `libopenweave.a`
    ESP-IDF 3.3 assumed that `lib(componentname).a` is built
- dns_getserver now returns `const ip_addr_t*`

Testing this PR requires https://github.com/openweave/openweave-esp32-demo/pull/14 and its lwip submodule should be checked out to https://github.com/openweave/openweave-esp32-lwip/pull/2

Tested features:
 - [x] Compiling
 - [x] Connect via BLE
 - [x] Network provisioning
   - [x] WiFi
 - [x] Pairing
 - [x] `reset-config` from WDM
 - [ ] *more tests pending*

### CODE REVIEW

File: build/esp32/components/openweave/component.mk [hunk 1]
```diff
@@ -174,7 +174,43 @@ install-weave : | build-weave
...
+	$(OUTPUT_DIR)/lib/libSystemLayer.d       \
+	$(OUTPUT_DIR)/lib/libuECC.d              \
+	$(OUTPUT_DIR)/lib/libWarm.d              \
+	$(OUTPUT_DIR)/lib/libDeviceLayer.d       \
+
+build-weave : configure-weave
+	echo "BUILD OPENWEAVE..."
+	MAKEFLAGS= make -C $(OUTPUT_DIR) --no-print-directory all
+
+install-weave : | build-weave
```
Comment (jaylogue) [inline build/esp32/components/openweave/component.mk:194]: Duplicate rule.

Comment (erjiaqing) [reply build/esp32/components/openweave/component.mk]: My fault, resolved.

File: build/esp32/components/openweave/component.mk [hunk 2]
```diff
@@ -174,7 +174,43 @@ install-weave : | build-weave
...
+	echo "INSTALL OPENWEAVE..."
+	MAKEFLAGS= make -C $(OUTPUT_DIR) --no-print-directory install
+
+# Always remove archives after built, the timestamp of the archive might not change
+# sometimes if not do so, then the image built will using outdated code.
+$(OUTPUT_DIR)/libopenweave.a : $(OPENWEAVE_STATIC_ARCHIVES)
```
[the remark this answers is not in the document]
Comment (erjiaqing) [reply build/esp32/components/openweave/component.mk]: I tried to rewrite this in the new commit, PTAL.

**Comment by erjiaqing** [comment]
I have added some compile time checks to let CI with esp-idf 3.0 happy, these can be removed if CI is upgraded to esp-idf 3.3

**Comment by lanyuwen** [comment]
> I have added some compile time checks to let CI with esp-idf 3.0 happy, these can be removed if CI is upgraded to esp-idf 3.3

I would suggest to focus on esp-idf 3.3 so that these compile time checks could be removed.

**Comment by erjiaqing** [comment]
> > I have added some compile time checks to let CI with esp-idf 3.0 happy, these can be removed if CI is upgraded to esp-idf 3.3
> 
> I would suggest to focus on esp-idf 3.3 so that these compile time checks could be removed.

True, I would suggest create 3 prs

1. Add esp-idf 3.3 support so both 3.0 and 3.3 can build
2. Upgrade CI to esp-idf 3.3 (or even newer?)
3. Drop esp-idf 3.0 support

**Comment by lanyuwen** [comment]
> > > I have added some compile time checks to let CI with esp-idf 3.0 happy, these can be removed if CI is upgraded to esp-idf 3.3
> > 
> > 
> > I would suggest to focus on esp-idf 3.3 so that these compile time checks could be removed.
> 
> True, I would suggest create 3 prs
> 
> 1. Add esp-idf 3.3 support so both 3.0 and 3.3 can build
> 2. Upgrade CI to esp-idf 3.3 (or even newer?)
> 3. Drop esp-idf 3.0 support

I think we can simply move on to support esp-idf 3.3 directly. We don't have to make CI happy all the time - especially when we want to upgrade the toolchains. 

@robszewczyk thoughts?

**Comment by kghost** [comment]
@erjiaqing Can you explain the detailed step to direct upgrade to esp-idf 3.3. How many repos and how many PRs are involved, are they all ready or not.

**Comment by erjiaqing** [comment]
@robszewczyk PTAL

**Pull request merged**